<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Load-data" data-toc-modified-id="Load-data-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Load data</a></span><ul class="toc-item"><li><span><a href="#Without-data-standardization" data-toc-modified-id="Without-data-standardization-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Without data standardization</a></span></li><li><span><a href="#Model-definition-(Vanilla-MLP)" data-toc-modified-id="Model-definition-(Vanilla-MLP)-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>Model definition (Vanilla MLP)</a></span></li></ul></li><li><span><a href="#Looking-at-the-model-structure" data-toc-modified-id="Looking-at-the-model-structure-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Looking at the model structure</a></span><ul class="toc-item"><li><span><a href="#Model-performance" data-toc-modified-id="Model-performance-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Model performance</a></span></li></ul></li><li><span><a href="#With-data-standardization" data-toc-modified-id="With-data-standardization-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>With data standardization</a></span><ul class="toc-item"><li><span><a href="#Model-performance" data-toc-modified-id="Model-performance-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Model performance</a></span></li></ul></li><li><span><a href="#MLP-with-tweaks" data-toc-modified-id="MLP-with-tweaks-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>MLP with tweaks</a></span><ul class="toc-item"><li><span><a href="#Model-performance" data-toc-modified-id="Model-performance-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Model performance</a></span></li><li><span><a href="#Training-metrics-changes" data-toc-modified-id="Training-metrics-changes-4.2"><span class="toc-item-num">4.2&nbsp;&nbsp;</span>Training metrics changes</a></span></li></ul></li><li><span><a href="#Deep-MLP-with-random-grid-search" data-toc-modified-id="Deep-MLP-with-random-grid-search-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Deep MLP with random grid search</a></span><ul class="toc-item"><li><span><a href="#Defining-the-grid-space" data-toc-modified-id="Defining-the-grid-space-5.1"><span class="toc-item-num">5.1&nbsp;&nbsp;</span>Defining the grid space</a></span></li><li><span><a href="#using-the-best-model-for-evaluation" data-toc-modified-id="using-the-best-model-for-evaluation-5.2"><span class="toc-item-num">5.2&nbsp;&nbsp;</span>using the best model for evaluation</a></span></li></ul></li></ul></div>

In [3]:
import sys

In [4]:
sys.path.append("c:\\users\\chait\\anaconda3\\lib\\site-packages")

In [5]:
import pandas as pd
import tensorflow.keras
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Dropout
from tensorflow.keras.optimizers import RMSprop

from tensorflow.keras.layers import BatchNormalization
from tensorflow.keras.callbacks import EarlyStopping, ModelCheckpoint
from tensorflow.keras.optimizers import Adam

from tensorflow.keras.wrappers.scikit_learn import KerasClassifier

import numpy as np
from sklearn.model_selection import RandomizedSearchCV

from tensorflow.keras import backend as K

def recall(y_true, y_pred):
    true_positives = K.sum(K.round(K.clip(y_true * y_pred, 0, 1)))
    possible_positives = K.sum(K.round(K.clip(y_true, 0, 1)))
    recall = true_positives / (possible_positives + K.epsilon())
    return recall

In [6]:
def convert_categorical_for_dl(x):
    if x =="ENGAGED":return 1 
    else: return 0

## Load data

In [7]:
X_train = pd.read_csv("model_data/X_train.csv")
y_train = pd.read_csv("model_data/y_train.csv")
X_validation = pd.read_csv("model_data/X_validation.csv")
y_validation = pd.read_csv("model_data/y_validation.csv")

In [8]:

batch_size = 128
num_classes = 2
epochs = 20

# the data, split between train and test sets

x_train = X_train.values
x_validation = X_validation.values


In [9]:
print(x_train.shape[0], 'train samples')

print(x_validation.shape[0], 'test samples')

12835 train samples
5502 test samples


In [10]:
X_train.shape[0]

12835

In [11]:
y_train = pd.Series(y_train.is_user_engaged).apply(lambda x : convert_categorical_for_dl(x))

In [12]:
y_validation = pd.Series(y_validation.is_user_engaged).apply(lambda x : convert_categorical_for_dl(x))

In [13]:
# convert class vectors to binary class matrices
y_train = tensorflow.keras.utils.to_categorical(y_train, num_classes)
y_validation = tensorflow.keras.utils.to_categorical(y_validation, num_classes)

### Without data standardization

### Model definition (Vanilla MLP)

In [14]:
model = Sequential()
model.add(Dense(512, activation='relu', input_shape=(X_train.shape[1],)))
model.add(Dropout(0.2))
model.add(Dense(512, activation='relu'))
model.add(Dropout(0.2))
model.add(Dense(num_classes, activation='softmax'))

model.summary()

Model: "sequential"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense (Dense)                (None, 512)               20992     
_________________________________________________________________
dropout (Dropout)            (None, 512)               0         
_________________________________________________________________
dense_1 (Dense)              (None, 512)               262656    
_________________________________________________________________
dropout_1 (Dropout)          (None, 512)               0         
_________________________________________________________________
dense_2 (Dense)              (None, 2)                 1026      
Total params: 284,674
Trainable params: 284,674
Non-trainable params: 0
_________________________________________________________________


## Looking at the model structure

In [15]:
model.summary()

Model: "sequential"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense (Dense)                (None, 512)               20992     
_________________________________________________________________
dropout (Dropout)            (None, 512)               0         
_________________________________________________________________
dense_1 (Dense)              (None, 512)               262656    
_________________________________________________________________
dropout_1 (Dropout)          (None, 512)               0         
_________________________________________________________________
dense_2 (Dense)              (None, 2)                 1026      
Total params: 284,674
Trainable params: 284,674
Non-trainable params: 0
_________________________________________________________________


In [16]:
x_train.shape

(12835, 40)

In [17]:
y_train.shape

(12835, 2)

In [18]:
model.compile(loss=['binary_crossentropy'],
              optimizer=RMSprop(),
              metrics=['accuracy', recall])

In [19]:
history = model.fit(x_train, y_train,
                    batch_size=batch_size,
                    epochs=epochs,
                    verbose=1,
                    validation_data=(x_validation, y_validation))
score = model.evaluate(x_validation, y_validation, verbose=0)
print('Validation Loss:', score[0])
print('Validation accuracy:', score[1])

Train on 12835 samples, validate on 5502 samples
Epoch 1/20
12835/12835 [==============================] - 4s 327us/sample - loss: 1.2267 - accuracy: 0.8644 - recall: 0.8652 - val_loss: 1.2193 - val_accuracy: 0.8753 - val_recall: 0.8753
Epoch 2/20
12835/12835 [==============================] - 1s 103us/sample - loss: 0.9691 - accuracy: 0.8778 - recall: 0.8772 - val_loss: 0.9713 - val_accuracy: 0.8566 - val_recall: 0.8566
Epoch 3/20
12835/12835 [==============================] - 1s 111us/sample - loss: 0.5997 - accuracy: 0.8811 - recall: 0.8805 - val_loss: 0.3631 - val_accuracy: 0.8844 - val_recall: 0.8844
Epoch 4/20
12835/12835 [==============================] - 2s 133us/sample - loss: 0.3780 - accuracy: 0.8889 - recall: 0.8881 - val_loss: 0.3133 - val_accuracy: 0.8811 - val_recall: 0.8811
Epoch 5/20
12835/12835 [==============================] - 2s 144us/sample - loss: 0.3289 - accuracy: 0.8918 - recall: 0.8917 - val_loss: 0.2795 - val_accuracy: 0.8951 - val_recall: 0.8951
Epoch 6/20


### Model performance

Deep learning model acheived a 90% recall on train and validation, no overfitting.
    - If we are looking at an interpretable model, we can use the decision tree model or random forest with some more hyper parameter tuning.
    - If we are looking for a model with high performance but no intepretability, the MLP model is a good choice.

## With data standardization

In [26]:
from sklearn.preprocessing import StandardScaler

In [27]:
sc = StandardScaler()

In [28]:
sc.fit(X_train)

StandardScaler(copy=True, with_mean=True, with_std=True)

In [29]:
x_train_std = sc.transform(X_train)
x_validation_std = sc.transform(X_validation)


In [30]:
model = Sequential()
model.add(Dense(512, activation='relu', input_shape=(X_train.shape[1],)))
model.add(Dropout(0.2))
model.add(Dense(512, activation='relu'))
model.add(Dropout(0.2))
model.add(Dense(num_classes, activation='softmax'))

model.summary()

Model: "sequential_1"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_3 (Dense)              (None, 512)               20992     
_________________________________________________________________
dropout_2 (Dropout)          (None, 512)               0         
_________________________________________________________________
dense_4 (Dense)              (None, 512)               262656    
_________________________________________________________________
dropout_3 (Dropout)          (None, 512)               0         
_________________________________________________________________
dense_5 (Dense)              (None, 2)                 1026      
Total params: 284,674
Trainable params: 284,674
Non-trainable params: 0
_________________________________________________________________


In [31]:
model.compile(loss=['binary_crossentropy'],
              optimizer=RMSprop(),
              metrics=['accuracy', recall])

In [32]:
history = model.fit(x_train_std, y_train,
                    batch_size=batch_size,
                    epochs=epochs,
                    verbose=1,
                    validation_data=(x_validation_std, y_validation))
score = model.evaluate(x_validation_std, y_validation, verbose=0)
print('Validation Loss:', score[0])
print('Validation accuracy:', score[1])

Train on 12835 samples, validate on 5502 samples
Epoch 1/20
12835/12835 [==============================] - 3s 213us/sample - loss: 0.2963 - accuracy: 0.8707 - recall: 0.8708 - val_loss: 0.2745 - val_accuracy: 0.8768 - val_recall: 0.8768
Epoch 2/20
12835/12835 [==============================] - 1s 109us/sample - loss: 0.2579 - accuracy: 0.8891 - recall: 0.8892 - val_loss: 0.2780 - val_accuracy: 0.8686 - val_recall: 0.8686
Epoch 3/20
12835/12835 [==============================] - 1s 108us/sample - loss: 0.2479 - accuracy: 0.8908 - recall: 0.8903 - val_loss: 0.2861 - val_accuracy: 0.8579 - val_recall: 0.8579
Epoch 4/20
12835/12835 [==============================] - 1s 103us/sample - loss: 0.2426 - accuracy: 0.8912 - recall: 0.8905 - val_loss: 0.2583 - val_accuracy: 0.8810 - val_recall: 0.8810
Epoch 5/20
12835/12835 [==============================] - 1s 101us/sample - loss: 0.2391 - accuracy: 0.8954 - recall: 0.8952 - val_loss: 0.2443 - val_accuracy: 0.8922 - val_recall: 0.8922
Epoch 6/20


### Model performance

The model performance with stardardized data is similar to that of the model built without standardization.
Recall of ~90% on both train and validation data

## MLP with tweaks

In [81]:
es = EarlyStopping(monitor="val_recall", patience=30, min_delta=0.5)
filepath="model_weights/weights-improvement-{epoch:02d}-recall{val_recall:.2f}.hdf5"
checkpoint = ModelCheckpoint(filepath, monitor='val_recall', verbose=1, save_best_only=True, mode='max')

In [82]:
from tensorflow.keras.callbacks import CSVLogger

csvl = CSVLogger(
    filename='training_logs/training.log',
    separator=',', 
    append=False)

In [83]:
callbacks = [es,checkpoint,csvl]

In [140]:
# del model

In [141]:
X_train.shape

(12835, 40)

In [142]:
model = Sequential()
model.add(Dense(512, activation='relu', input_shape=(X_train.shape[1],)))
model.add(Dropout(0.2))
model.add(Dense(512, activation='relu'))
model.add(Dropout(0.2))
model.add(Dense(num_classes, activation='sigmoid'))
model.summary()

Model: "sequential_11"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_33 (Dense)             (None, 512)               20992     
_________________________________________________________________
dropout_22 (Dropout)         (None, 512)               0         
_________________________________________________________________
dense_34 (Dense)             (None, 512)               262656    
_________________________________________________________________
dropout_23 (Dropout)         (None, 512)               0         
_________________________________________________________________
dense_35 (Dense)             (None, 2)                 1026      
Total params: 284,674
Trainable params: 284,674
Non-trainable params: 0
_________________________________________________________________


In [143]:
epochs = 1000

In [144]:
batch_size = 64

In [145]:
model.compile(loss=['binary_crossentropy'],
              optimizer=Adam(),
              metrics=['accuracy', recall])
history = model.fit(x_train, y_train,
                    batch_size=batch_size,
                    epochs=epochs,
                    verbose=1,
                    validation_data=(x_validation, y_validation),callbacks = callbacks)

Train on 12835 samples, validate on 5502 samples
Epoch 1/1000
12736/12835 [============================>.] - ETA: 0s - loss: 17.6941 - accuracy: 0.8452 - recall: 0.8449
Epoch 00001: val_recall did not improve from 0.90549
12835/12835 [==============================] - 3s 197us/sample - loss: 17.5660 - accuracy: 0.8455 - recall: 0.8452 - val_loss: 9.3523 - val_accuracy: 0.8669 - val_recall: 0.8773
Epoch 2/1000
12544/12835 [============================>.] - ETA: 0s - loss: 17.1161 - accuracy: 0.8509 - recall: 0.8522
Epoch 00002: val_recall did not improve from 0.90549
12835/12835 [==============================] - 2s 141us/sample - loss: 16.7958 - accuracy: 0.8512 - recall: 0.8524 - val_loss: 29.3576 - val_accuracy: 0.8848 - val_recall: 0.8788
Epoch 3/1000
12800/12835 [============================>.] - ETA: 0s - loss: 5.0840 - accuracy: 0.8446 - recall: 0.8397
Epoch 00003: val_recall did not improve from 0.90549
12835/12835 [==============================] - 2s 130us/sample - loss: 5.071

Epoch 24/1000
12800/12835 [============================>.] - ETA: 0s - loss: 0.3090 - accuracy: 0.9000 - recall: 0.8995
Epoch 00024: val_recall did not improve from 0.90822
12835/12835 [==============================] - 2s 129us/sample - loss: 0.3091 - accuracy: 0.8999 - recall: 0.8995 - val_loss: 8.7722 - val_accuracy: 0.8971 - val_recall: 0.8971
Epoch 25/1000
12544/12835 [============================>.] - ETA: 0s - loss: 0.2674 - accuracy: 0.8984 - recall: 0.8996
Epoch 00025: val_recall did not improve from 0.90822
12835/12835 [==============================] - 2s 132us/sample - loss: 0.2679 - accuracy: 0.8982 - recall: 0.8994 - val_loss: 8.6760 - val_accuracy: 0.8984 - val_recall: 0.8973
Epoch 26/1000
12672/12835 [============================>.] - ETA: 0s - loss: 0.2749 - accuracy: 0.8982 - recall: 0.8984
Epoch 00026: val_recall did not improve from 0.90822
12835/12835 [==============================] - 2s 131us/sample - loss: 0.2741 - accuracy: 0.8983 - recall: 0.8986 - val_loss: 7

In [146]:
score = model.evaluate(x_validation, y_validation, verbose=0)
print('Validation Loss:', score[0])
print('Validation accuracy:', score[1])
print('Validation Recall:', score[2])

Validation Loss: 8.28888676851522
Validation accuracy: 0.89921844
Validation Recall: 0.89914


In [147]:
score

[8.28888676851522, 0.89921844, 0.89914]

### Model performance

The validation and train recall are ~90% I have only done a dropout and early stopping, building a deeper netwrok may improve the recall.

### Training metrics changes

In [148]:
import pandas as pd

In [149]:
logs = pd.read_csv("training_logs/training.log")

In [150]:
logs.head()

epoch  accuracy         loss    recall  val_accuracy    val_loss  \
0      0  0.845462    17.566042  0.845229      0.866867    9.352285   
1      1  0.851227    16.795758  0.852419      0.884769   29.357563   
2      2  0.844877     5.071482  0.840201      0.860051   27.594191   
3      3  0.862213   320.561889  0.863886      0.887586  104.754415   
4      4  0.849552  2659.429828  0.855930      0.874409  159.011708   

   val_recall  
0    0.877321  
1    0.878763  
2    0.868782  
3    0.895308  
4    0.908219

In [151]:
import plotly
import plotly.offline as pyoff
import plotly.figure_factory as ff
from plotly.offline import init_notebook_mode, iplot, plot
import plotly.graph_objs as go
init_notebook_mode(connected = True)

In [152]:
data  = []
for i in logs.columns:
    if i != "epoch":
        data.append(go.Scatter(x=logs.epoch, y=logs[i], name = i))

layout = go.Layout(title ="Training loss and accuracy changes")        
fig = go.Figure(data = data, layout= layout)

fig.show()

## Deep MLP with random grid search

In [16]:
es = EarlyStopping(monitor="val_recall", patience=10, min_delta=0.5)
filepath="model_weights/weights-improvement-gs-{epoch:02d}-recall{val_recall:.2f}.hdf5"
checkpoint = ModelCheckpoint(filepath, monitor='val_recall', verbose=1, save_best_only=True, mode='max')

from tensorflow.keras.callbacks import CSVLogger

csvl = CSVLogger(
    filename='training_logs/grid_search.log',
    separator=',', 
    append=False)

callbacks = [es,checkpoint,csvl]

In [33]:
def build_model(optimizer='adam', dense_dims1=512,dense_dims2 = 128, dense_dims3 = 64, 
               dropout1 = 0.2, dropout2 = 0.2, dropout3 = 0.2):
    model = Sequential()
    model.add(Dense(dense_dims1, activation='relu', input_dim=X_train.shape[1]))
    model.add(Dropout(dropout1))
    model.add(Dense(dense_dims2, activation='relu'))
    model.add(Dropout(dropout2))
    model.add(Dense(dense_dims3, activation='relu'))
    model.add(Dropout(dropout3))
    model.add(Dense(2, activation='sigmoid'))
    model.compile(optimizer=optimizer,
                  loss='binary_crossentropy',
                  metrics=['accuracy', recall])
    return model

keras_classifier = KerasClassifier(build_model, epochs=2)

### Defining the grid space

In [34]:
n_epochs = [1000]
dense_dims1 = [128,512]
dense_dims2 = [128,512]
dense_dims3 = [128,512]
dropout1 = np.linspace(0,0.4,2)
dropout2 = np.linspace(0,0.4,2)
dropout3 = np.linspace(0,0.4,2)

In [35]:
gs = RandomizedSearchCV(keras_classifier, {'epochs': n_epochs, 'dense_dims1':dense_dims1,
                                     'dense_dims2':dense_dims2,
                                     'dense_dims3':dense_dims3,
                                     
                                    'dropout1' :dropout1,
                                    'dropout2' :dropout2,
                                    'dropout3' :dropout3})

In [36]:
gs.param_distributions

{'epochs': [1000],
 'dense_dims1': [128, 512],
 'dense_dims2': [128, 512],
 'dense_dims3': [128, 512],
 'dropout1': array([0. , 0.4]),
 'dropout2': array([0. , 0.4]),
 'dropout3': array([0. , 0.4])}

In [37]:
%%time
gs.fit(X_train,y_train,verbose=1,
        validation_data=(x_validation, y_validation),callbacks = callbacks)

Train on 10268 samples, validate on 5502 samples
Epoch 1/1000
10112/10268 [============================>.] - ETA: 0s - loss: 0.8717 - accuracy: 0.8690 - recall: 0.8659
Epoch 00001: val_recall did not improve from 0.92422
10268/10268 [==============================] - 3s 295us/sample - loss: 0.8693 - accuracy: 0.8694 - recall: 0.8659 - val_loss: 0.6733 - val_accuracy: 0.8808 - val_recall: 0.8352
Epoch 2/1000
10016/10268 [============================>.] - ETA: 0s - loss: 0.4256 - accuracy: 0.8865 - recall: 0.8871
Epoch 00002: val_recall did not improve from 0.92422
10268/10268 [==============================] - 3s 254us/sample - loss: 0.4223 - accuracy: 0.8867 - recall: 0.8876 - val_loss: 0.2985 - val_accuracy: 0.8831 - val_recall: 0.8880
Epoch 3/1000
10016/10268 [============================>.] - ETA: 0s - loss: 0.3108 - accuracy: 0.8965 - recall: 0.8968
Epoch 00003: val_recall did not improve from 0.92422
10268/10268 [==============================] - 3s 250us/sample - loss: 0.3102 - a

2567/2567 [==============================] - 0s 108us/sample - loss: 0.2496 - accuracy: 0.9038 - recall: 0.9029


Train on 10268 samples, validate on 5502 samples
Epoch 1/1000
10016/10268 [============================>.] - ETA: 0s - loss: 0.9060 - accuracy: 0.8563 - recall: 0.8558
Epoch 00001: val_recall improved from 0.92422 to 0.93840, saving model to model_weights/weights-improvement-gs-01-recall0.94.hdf5
10268/10268 [==============================] - 3s 277us/sample - loss: 0.8994 - accuracy: 0.8563 - recall: 0.8556 - val_loss: 0.7032 - val_accuracy: 0.8625 - val_recall: 0.9384
Epoch 2/1000
10048/10268 [============================>.] - ETA: 0s - loss: 0.4381 - accuracy: 0.8807 - recall: 0.8830
Epoch 00002: val_recall did not improve from 0.93840
10268/10268 [==============================] - 3s 245us/sample - loss: 0.4372 - accuracy: 0.8808 - recall: 0.8831 - val_loss: 0.3315 - val_accuracy: 0.8919 - val_recall: 0.8971
Epoch 3/1000
10176/10268 [============================>.] - ETA: 0s - loss: 0.2947 - accuracy: 0.8946 - recall: 0.8937
Epoch 00003: val_recall did not improve from 0.93840
1026

2567/2567 [==============================] - 0s 108us/sample - loss: 0.2313 - accuracy: 0.9001 - recall: 0.9012


Train on 10268 samples, validate on 5502 samples
Epoch 1/1000
 9984/10268 [============================>.] - ETA: 0s - loss: 0.8180 - accuracy: 0.8695 - recall: 0.8717
Epoch 00001: val_recall did not improve from 0.93840
10268/10268 [==============================] - 3s 297us/sample - loss: 0.8041 - accuracy: 0.8703 - recall: 0.8730 - val_loss: 0.3899 - val_accuracy: 0.8764 - val_recall: 0.8702
Epoch 2/1000
10208/10268 [============================>.] - ETA: 0s - loss: 0.4250 - accuracy: 0.8900 - recall: 0.8920
Epoch 00002: val_recall did not improve from 0.93840
10268/10268 [==============================] - 3s 337us/sample - loss: 0.4247 - accuracy: 0.8900 - recall: 0.8921 - val_loss: 0.3135 - val_accuracy: 0.8872 - val_recall: 0.8981
Epoch 3/1000
10240/10268 [============================>.] - ETA: 0s - loss: 0.2737 - accuracy: 0.8988 - recall: 0.8978
Epoch 00003: val_recall did not improve from 0.93840
10268/10268 [==============================] - 3s 302us/sample - loss: 0.2748 - a

2567/2567 [==============================] - 0s 96us/sample - loss: 0.2463 - accuracy: 0.8954 - recall: 0.8951


Train on 10268 samples, validate on 5502 samples
Epoch 1/1000
10080/10268 [============================>.] - ETA: 0s - loss: 0.7365 - accuracy: 0.8705 - recall: 0.8742
Epoch 00001: val_recall did not improve from 0.93840
10268/10268 [==============================] - 3s 272us/sample - loss: 0.7286 - accuracy: 0.8713 - recall: 0.8747 - val_loss: 0.4312 - val_accuracy: 0.8850 - val_recall: 0.8888
Epoch 2/1000
10208/10268 [============================>.] - ETA: 0s - loss: 0.3204 - accuracy: 0.8900 - recall: 0.8901
Epoch 00002: val_recall did not improve from 0.93840
10268/10268 [==============================] - 2s 238us/sample - loss: 0.3218 - accuracy: 0.8892 - recall: 0.8893 - val_loss: 0.2820 - val_accuracy: 0.8886 - val_recall: 0.8980
Epoch 3/1000
10176/10268 [============================>.] - ETA: 0s - loss: 0.3018 - accuracy: 0.8963 - recall: 0.8978
Epoch 00003: val_recall did not improve from 0.93840
10268/10268 [==============================] - 3s 261us/sample - loss: 0.3025 - a

2567/2567 [==============================] - 0s 110us/sample - loss: 0.2490 - accuracy: 0.9100 - recall: 0.9083


Train on 10268 samples, validate on 5502 samples
Epoch 1/1000
10048/10268 [============================>.] - ETA: 0s - loss: 0.8775 - accuracy: 0.8598 - recall: 0.8544
Epoch 00001: val_recall did not improve from 0.93840
10268/10268 [==============================] - 3s 305us/sample - loss: 0.8678 - accuracy: 0.8602 - recall: 0.8553 - val_loss: 0.4515 - val_accuracy: 0.8859 - val_recall: 0.8720
Epoch 2/1000
10144/10268 [============================>.] - ETA: 0s - loss: 0.3744 - accuracy: 0.8833 - recall: 0.8795
Epoch 00002: val_recall did not improve from 0.93840
10268/10268 [==============================] - 3s 260us/sample - loss: 0.3728 - accuracy: 0.8834 - recall: 0.8797 - val_loss: 0.3062 - val_accuracy: 0.8866 - val_recall: 0.8933
Epoch 3/1000
10144/10268 [============================>.] - ETA: 0s - loss: 0.2909 - accuracy: 0.8959 - recall: 0.8988
Epoch 00003: val_recall did not improve from 0.93840
10268/10268 [==============================] - 3s 259us/sample - loss: 0.2931 - a

2567/2567 [==============================] - 0s 123us/sample - loss: 0.2206 - accuracy: 0.9110 - recall: 0.9117


Train on 10268 samples, validate on 5502 samples
Epoch 1/1000
10144/10268 [============================>.] - ETA: 0s - loss: 0.7850 - accuracy: 0.8609 - recall: 0.8559
Epoch 00001: val_recall did not improve from 0.93840
10268/10268 [==============================] - 3s 320us/sample - loss: 0.7795 - accuracy: 0.8607 - recall: 0.8555 - val_loss: 0.3645 - val_accuracy: 0.8798 - val_recall: 0.8769
Epoch 2/1000
10208/10268 [============================>.] - ETA: 0s - loss: 0.3872 - accuracy: 0.8848 - recall: 0.8838
Epoch 00002: val_recall did not improve from 0.93840
10268/10268 [==============================] - 3s 272us/sample - loss: 0.3876 - accuracy: 0.8850 - recall: 0.8839 - val_loss: 0.2842 - val_accuracy: 0.8880 - val_recall: 0.8860
Epoch 3/1000
10080/10268 [============================>.] - ETA: 0s - loss: 0.3014 - accuracy: 0.8953 - recall: 0.8954
Epoch 00003: val_recall did not improve from 0.93840
10268/10268 [==============================] - 3s 300us/sample - loss: 0.3001 - a

2567/2567 [==============================] - 0s 155us/sample - loss: 0.2343 - accuracy: 0.9049 - recall: 0.9064


Train on 10268 samples, validate on 5502 samples
Epoch 1/1000
10080/10268 [============================>.] - ETA: 0s - loss: 0.8372 - accuracy: 0.8621 - recall: 0.8627
Epoch 00001: val_recall did not improve from 0.93840
10268/10268 [==============================] - 3s 281us/sample - loss: 0.8280 - accuracy: 0.8621 - recall: 0.8629 - val_loss: 0.4209 - val_accuracy: 0.8828 - val_recall: 0.8977
Epoch 2/1000
10176/10268 [============================>.] - ETA: 0s - loss: 0.3652 - accuracy: 0.8844 - recall: 0.8839- ETA: 1s - loss: 0.4640 - 
Epoch 00002: val_recall did not improve from 0.93840
10268/10268 [==============================] - 3s 260us/sample - loss: 0.3636 - accuracy: 0.8848 - recall: 0.8844 - val_loss: 0.2697 - val_accuracy: 0.8916 - val_recall: 0.8924
Epoch 3/1000
10144/10268 [============================>.] - ETA: 0s - loss: 0.2957 - accuracy: 0.8962 - recall: 0.8971- ETA: 1s - loss: 0.3294 - 
Epoch 00003: val_recall did not improve from 0.93840
10268/10268 [==============

2567/2567 [==============================] - 0s 100us/sample - loss: 0.2279 - accuracy: 0.9014 - recall: 0.9001


Train on 10268 samples, validate on 5502 samples
Epoch 1/1000
10240/10268 [============================>.] - ETA: 0s - loss: 0.8319 - accuracy: 0.8673 - recall: 0.8679
Epoch 00001: val_recall did not improve from 0.93840
10268/10268 [==============================] - 3s 298us/sample - loss: 0.8307 - accuracy: 0.8674 - recall: 0.8682 - val_loss: 0.4538 - val_accuracy: 0.8818 - val_recall: 0.8806
Epoch 2/1000
10016/10268 [============================>.] - ETA: 0s - loss: 0.4089 - accuracy: 0.8819 - recall: 0.8833
Epoch 00002: val_recall did not improve from 0.93840
10268/10268 [==============================] - 3s 266us/sample - loss: 0.4056 - accuracy: 0.8827 - recall: 0.8842 - val_loss: 0.2849 - val_accuracy: 0.8898 - val_recall: 0.8961
Epoch 3/1000
10048/10268 [============================>.] - ETA: 0s - loss: 0.3181 - accuracy: 0.8947 - recall: 0.8951
Epoch 00003: val_recall did not improve from 0.93840
10268/10268 [==============================] - 3s 300us/sample - loss: 0.3171 - a

2567/2567 [==============================] - 0s 101us/sample - loss: 0.2525 - accuracy: 0.8923 - recall: 0.8925


Train on 10268 samples, validate on 5502 samples
Epoch 1/1000
10240/10268 [============================>.] - ETA: 0s - loss: 0.8087 - accuracy: 0.8633 - recall: 0.8653
Epoch 00001: val_recall did not improve from 0.93840
10268/10268 [==============================] - 3s 279us/sample - loss: 0.8073 - accuracy: 0.8634 - recall: 0.8654 - val_loss: 0.4678 - val_accuracy: 0.8632 - val_recall: 0.8722
Epoch 2/1000
10048/10268 [============================>.] - ETA: 0s - loss: 0.4038 - accuracy: 0.8811 - recall: 0.8808
Epoch 00002: val_recall did not improve from 0.93840
10268/10268 [==============================] - 3s 254us/sample - loss: 0.4010 - accuracy: 0.8822 - recall: 0.8820 - val_loss: 0.3117 - val_accuracy: 0.8844 - val_recall: 0.8864
Epoch 3/1000
10176/10268 [============================>.] - ETA: 0s - loss: 0.3072 - accuracy: 0.8911 - recall: 0.8917
Epoch 00003: val_recall did not improve from 0.93840
10268/10268 [==============================] - 3s 247us/sample - loss: 0.3076 - a

2567/2567 [==============================] - 0s 102us/sample - loss: 0.2479 - accuracy: 0.9083 - recall: 0.9080


Train on 10268 samples, validate on 5502 samples
Epoch 1/1000
10048/10268 [============================>.] - ETA: 0s - loss: 0.8545 - accuracy: 0.8573 - recall: 0.8581
Epoch 00001: val_recall did not improve from 0.93840
10268/10268 [==============================] - 3s 278us/sample - loss: 0.8471 - accuracy: 0.8579 - recall: 0.8584 - val_loss: 0.4295 - val_accuracy: 0.8841 - val_recall: 0.8637
Epoch 2/1000
10080/10268 [============================>.] - ETA: 0s - loss: 0.4147 - accuracy: 0.8802 - recall: 0.8770
Epoch 00002: val_recall did not improve from 0.93840
10268/10268 [==============================] - 3s 246us/sample - loss: 0.4138 - accuracy: 0.8804 - recall: 0.8772 - val_loss: 0.3747 - val_accuracy: 0.8909 - val_recall: 0.8868
Epoch 3/1000
10080/10268 [============================>.] - ETA: 0s - loss: 0.3391 - accuracy: 0.8907 - recall: 0.8903
Epoch 00003: val_recall did not improve from 0.93840
10268/10268 [==============================] - 3s 245us/sample - loss: 0.3390 - a

2567/2567 [==============================] - 0s 103us/sample - loss: 0.2145 - accuracy: 0.9131 - recall: 0.9147


Train on 10268 samples, validate on 5502 samples
Epoch 1/1000
10144/10268 [============================>.] - ETA: 0s - loss: 0.5625 - accuracy: 0.8684 - recall: 0.8664
Epoch 00001: val_recall did not improve from 0.93840
10268/10268 [==============================] - 2s 243us/sample - loss: 0.5604 - accuracy: 0.8681 - recall: 0.8662 - val_loss: 0.3586 - val_accuracy: 0.8683 - val_recall: 0.8686
Epoch 2/1000
10208/10268 [============================>.] - ETA: 0s - loss: 0.3250 - accuracy: 0.8858 - recall: 0.8910
Epoch 00002: val_recall did not improve from 0.93840
10268/10268 [==============================] - 2s 214us/sample - loss: 0.3253 - accuracy: 0.8856 - recall: 0.8907 - val_loss: 0.2742 - val_accuracy: 0.8876 - val_recall: 0.8932
Epoch 3/1000
10208/10268 [============================>.] - ETA: 0s - loss: 0.2927 - accuracy: 0.8892 - recall: 0.8912
Epoch 00003: val_recall did not improve from 0.93840
10268/10268 [==============================] - 2s 216us/sample - loss: 0.2918 - a

2567/2567 [==============================] - 0s 96us/sample - loss: 0.2444 - accuracy: 0.9028 - recall: 0.9022


Train on 10268 samples, validate on 5502 samples
Epoch 1/1000
10048/10268 [============================>.] - ETA: 0s - loss: 0.6578 - accuracy: 0.8666 - recall: 0.8681
Epoch 00001: val_recall did not improve from 0.93840
10268/10268 [==============================] - 3s 263us/sample - loss: 0.6555 - accuracy: 0.8664 - recall: 0.8679 - val_loss: 0.2944 - val_accuracy: 0.8849 - val_recall: 0.8868
Epoch 2/1000
10176/10268 [============================>.] - ETA: 0s - loss: 0.3279 - accuracy: 0.8858 - recall: 0.8876
Epoch 00002: val_recall did not improve from 0.93840
10268/10268 [==============================] - 2s 221us/sample - loss: 0.3270 - accuracy: 0.8861 - recall: 0.8879 - val_loss: 0.2713 - val_accuracy: 0.8885 - val_recall: 0.8886
Epoch 3/1000
10208/10268 [============================>.] - ETA: 0s - loss: 0.2959 - accuracy: 0.8892 - recall: 0.8899
Epoch 00003: val_recall did not improve from 0.93840
10268/10268 [==============================] - 2s 225us/sample - loss: 0.2957 - a

2567/2567 [==============================] - 0s 99us/sample - loss: 0.2366 - accuracy: 0.9012 - recall: 0.9024


Train on 10268 samples, validate on 5502 samples
Epoch 1/1000
10208/10268 [============================>.] - ETA: 0s - loss: 0.6096 - accuracy: 0.8696 - recall: 0.8719
Epoch 00001: val_recall did not improve from 0.93840
10268/10268 [==============================] - 3s 258us/sample - loss: 0.6079 - accuracy: 0.8696 - recall: 0.8719 - val_loss: 0.2879 - val_accuracy: 0.8793 - val_recall: 0.8804
Epoch 2/1000
 9920/10268 [===========================>..] - ETA: 0s - loss: 0.3496 - accuracy: 0.8879 - recall: 0.8854
Epoch 00002: val_recall did not improve from 0.93840
10268/10268 [==============================] - 2s 212us/sample - loss: 0.3493 - accuracy: 0.8872 - recall: 0.8846 - val_loss: 0.3555 - val_accuracy: 0.8869 - val_recall: 0.8839
Epoch 3/1000
10240/10268 [============================>.] - ETA: 0s - loss: 0.3034 - accuracy: 0.8910 - recall: 0.8890- ETA: 0s - loss: 0.3071 - accuracy: 0.8919 - 
Epoch 00003: val_recall did not improve from 0.93840
10268/10268 [======================

2567/2567 [==============================] - 0s 99us/sample - loss: 0.2509 - accuracy: 0.8921 - recall: 0.8914


Train on 10268 samples, validate on 5502 samples
Epoch 1/1000
10112/10268 [============================>.] - ETA: 0s - loss: 0.5508 - accuracy: 0.8713 - recall: 0.8688
Epoch 00001: val_recall did not improve from 0.93840
10268/10268 [==============================] - 3s 256us/sample - loss: 0.5469 - accuracy: 0.8719 - recall: 0.8695 - val_loss: 0.3535 - val_accuracy: 0.8732 - val_recall: 0.8750
Epoch 2/1000
10048/10268 [============================>.] - ETA: 0s - loss: 0.3241 - accuracy: 0.8890 - recall: 0.8888
Epoch 00002: val_recall did not improve from 0.93840
10268/10268 [==============================] - 3s 264us/sample - loss: 0.3231 - accuracy: 0.8898 - recall: 0.8896 - val_loss: 0.3122 - val_accuracy: 0.8810 - val_recall: 0.8811
Epoch 3/1000
10016/10268 [============================>.] - ETA: 0s - loss: 0.2938 - accuracy: 0.8942 - recall: 0.8941
Epoch 00003: val_recall did not improve from 0.93840
10268/10268 [==============================] - 2s 229us/sample - loss: 0.2943 - a

2567/2567 [==============================] - 0s 121us/sample - loss: 0.2970 - accuracy: 0.8933 - recall: 0.8939


Train on 10268 samples, validate on 5502 samples
Epoch 1/1000
10112/10268 [============================>.] - ETA: 0s - loss: 0.6676 - accuracy: 0.8544 - recall: 0.8511
Epoch 00001: val_recall did not improve from 0.93840
10268/10268 [==============================] - 3s 298us/sample - loss: 0.6652 - accuracy: 0.8537 - recall: 0.8505 - val_loss: 0.3333 - val_accuracy: 0.8550 - val_recall: 0.8551
Epoch 2/1000
10176/10268 [============================>.] - ETA: 0s - loss: 0.3399 - accuracy: 0.8602 - recall: 0.8601
Epoch 00002: val_recall did not improve from 0.93840
10268/10268 [==============================] - 3s 253us/sample - loss: 0.3395 - accuracy: 0.8602 - recall: 0.8600 - val_loss: 0.3107 - val_accuracy: 0.8550 - val_recall: 0.8549
Epoch 3/1000
10080/10268 [============================>.] - ETA: 0s - loss: 0.2988 - accuracy: 0.8691 - recall: 0.8804
Epoch 00003: val_recall did not improve from 0.93840
10268/10268 [==============================] - 2s 233us/sample - loss: 0.2986 - a

2567/2567 [==============================] - 0s 119us/sample - loss: 0.2252 - accuracy: 0.9122 - recall: 0.9124


Train on 10268 samples, validate on 5502 samples
Epoch 1/1000
10176/10268 [============================>.] - ETA: 0s - loss: 0.9425 - accuracy: 0.8528 - recall: 0.8541
Epoch 00001: val_recall did not improve from 0.93840
10268/10268 [==============================] - 3s 276us/sample - loss: 0.9372 - accuracy: 0.8531 - recall: 0.8542 - val_loss: 0.6158 - val_accuracy: 0.8591 - val_recall: 0.8575
Epoch 2/1000
10112/10268 [============================>.] - ETA: 0s - loss: 0.5517 - accuracy: 0.8598 - recall: 0.8576
Epoch 00002: val_recall did not improve from 0.93840
10268/10268 [==============================] - 2s 241us/sample - loss: 0.5513 - accuracy: 0.8596 - recall: 0.8573 - val_loss: 0.3820 - val_accuracy: 0.8741 - val_recall: 0.8710
Epoch 3/1000
 9984/10268 [============================>.] - ETA: 0s - loss: 0.4512 - accuracy: 0.8683 - recall: 0.8697
Epoch 00003: val_recall did not improve from 0.93840
10268/10268 [==============================] - 3s 260us/sample - loss: 0.4475 - a

2567/2567 [==============================] - 0s 125us/sample - loss: 0.2630 - accuracy: 0.8985 - recall: 0.8979


Train on 10268 samples, validate on 5502 samples
Epoch 1/1000
 9984/10268 [============================>.] - ETA: 0s - loss: 0.9427 - accuracy: 0.8444 - recall: 0.8402
Epoch 00001: val_recall did not improve from 0.93840
10268/10268 [==============================] - 3s 296us/sample - loss: 0.9329 - accuracy: 0.8445 - recall: 0.8407 - val_loss: 0.4434 - val_accuracy: 0.8739 - val_recall: 0.9006
Epoch 2/1000
10208/10268 [============================>.] - ETA: 0s - loss: 0.4901 - accuracy: 0.8611 - recall: 0.8661
Epoch 00002: val_recall did not improve from 0.93840
10268/10268 [==============================] - 3s 264us/sample - loss: 0.4898 - accuracy: 0.8613 - recall: 0.8662 - val_loss: 0.3476 - val_accuracy: 0.8753 - val_recall: 0.8708
Epoch 3/1000
10176/10268 [============================>.] - ETA: 0s - loss: 0.4134 - accuracy: 0.8706 - recall: 0.8701
Epoch 00003: val_recall did not improve from 0.93840
10268/10268 [==============================] - 3s 266us/sample - loss: 0.4133 - a

2567/2567 [==============================] - 0s 139us/sample - loss: 0.2503 - accuracy: 0.8991 - recall: 0.8974


Train on 10268 samples, validate on 5502 samples
Epoch 1/1000
10208/10268 [============================>.] - ETA: 0s - loss: 0.9120 - accuracy: 0.8474 - recall: 0.8469
Epoch 00001: val_recall improved from 0.93840 to 0.94130, saving model to model_weights/weights-improvement-gs-01-recall0.94.hdf5
10268/10268 [==============================] - 3s 286us/sample - loss: 0.9100 - accuracy: 0.8473 - recall: 0.8471 - val_loss: 0.4313 - val_accuracy: 0.8488 - val_recall: 0.9413
Epoch 2/1000
10016/10268 [============================>.] - ETA: 0s - loss: 0.4778 - accuracy: 0.8596 - recall: 0.8635
Epoch 00002: val_recall did not improve from 0.94130
10268/10268 [==============================] - 2s 239us/sample - loss: 0.4795 - accuracy: 0.8592 - recall: 0.8630 - val_loss: 0.3391 - val_accuracy: 0.8690 - val_recall: 0.8749
Epoch 3/1000
10240/10268 [============================>.] - ETA: 0s - loss: 0.4007 - accuracy: 0.8718 - recall: 0.8736
Epoch 00003: val_recall did not improve from 0.94130
1026

2567/2567 [==============================] - 0s 113us/sample - loss: 0.2870 - accuracy: 0.8894 - recall: 0.8871


Train on 10268 samples, validate on 5502 samples
Epoch 1/1000
10176/10268 [============================>.] - ETA: 0s - loss: 0.9589 - accuracy: 0.8518 - recall: 0.8567
Epoch 00001: val_recall did not improve from 0.94130
10268/10268 [==============================] - 4s 344us/sample - loss: 0.9548 - accuracy: 0.8519 - recall: 0.8572 - val_loss: 0.4501 - val_accuracy: 0.8656 - val_recall: 0.8956
Epoch 2/1000
10112/10268 [============================>.] - ETA: 0s - loss: 0.5007 - accuracy: 0.8538 - recall: 0.8500
Epoch 00002: val_recall did not improve from 0.94130
10268/10268 [==============================] - 3s 253us/sample - loss: 0.4992 - accuracy: 0.8542 - recall: 0.8503 - val_loss: 0.3270 - val_accuracy: 0.8577 - val_recall: 0.8573
Epoch 3/1000
10016/10268 [============================>.] - ETA: 0s - loss: 0.3871 - accuracy: 0.8749 - recall: 0.8765
Epoch 00003: val_recall did not improve from 0.94130
10268/10268 [==============================] - 2s 240us/sample - loss: 0.3884 - a

2567/2567 [==============================] - 0s 143us/sample - loss: 0.2767 - accuracy: 0.8927 - recall: 0.8927


Train on 10268 samples, validate on 5502 samples
Epoch 1/1000
10144/10268 [============================>.] - ETA: 0s - loss: 0.9984 - accuracy: 0.8494 - recall: 0.8536
Epoch 00001: val_recall did not improve from 0.94130
10268/10268 [==============================] - 3s 319us/sample - loss: 0.9943 - accuracy: 0.8495 - recall: 0.8538 - val_loss: 0.5081 - val_accuracy: 0.8633 - val_recall: 0.8433
Epoch 2/1000
10016/10268 [============================>.] - ETA: 0s - loss: 0.5429 - accuracy: 0.8493 - recall: 0.8329
Epoch 00002: val_recall did not improve from 0.94130
10268/10268 [==============================] - 3s 257us/sample - loss: 0.5407 - accuracy: 0.8499 - recall: 0.8338 - val_loss: 0.3648 - val_accuracy: 0.8550 - val_recall: 0.8550
Epoch 3/1000
10048/10268 [============================>.] - ETA: 0s - loss: 0.4066 - accuracy: 0.8654 - recall: 0.8642
Epoch 00003: val_recall did not improve from 0.94130
10268/10268 [==============================] - 3s 247us/sample - loss: 0.4050 - a

2567/2567 [==============================] - 0s 118us/sample - loss: 0.2398 - accuracy: 0.9129 - recall: 0.9151


Train on 10268 samples, validate on 5502 samples
Epoch 1/1000
10208/10268 [============================>.] - ETA: 0s - loss: 0.5243 - accuracy: 0.8672 - recall: 0.8654
Epoch 00001: val_recall did not improve from 0.94130
10268/10268 [==============================] - 3s 302us/sample - loss: 0.5242 - accuracy: 0.8671 - recall: 0.8651 - val_loss: 0.3168 - val_accuracy: 0.8847 - val_recall: 0.8917
Epoch 2/1000
10080/10268 [============================>.] - ETA: 0s - loss: 0.3521 - accuracy: 0.8869 - recall: 0.8865
Epoch 00002: val_recall did not improve from 0.94130
10268/10268 [==============================] - 3s 271us/sample - loss: 0.3501 - accuracy: 0.8871 - recall: 0.8868 - val_loss: 0.2923 - val_accuracy: 0.8872 - val_recall: 0.8886
Epoch 3/1000
10016/10268 [============================>.] - ETA: 0s - loss: 0.3064 - accuracy: 0.8872 - recall: 0.8872
Epoch 00003: val_recall did not improve from 0.94130
10268/10268 [==============================] - 2s 240us/sample - loss: 0.3051 - a

2567/2567 [==============================] - 0s 110us/sample - loss: 0.2500 - accuracy: 0.9036 - recall: 0.9068


Train on 10268 samples, validate on 5502 samples
Epoch 1/1000
10080/10268 [============================>.] - ETA: 0s - loss: 0.5340 - accuracy: 0.8676 - recall: 0.8699
Epoch 00001: val_recall did not improve from 0.94130
10268/10268 [==============================] - 3s 297us/sample - loss: 0.5315 - accuracy: 0.8676 - recall: 0.8699 - val_loss: 0.3479 - val_accuracy: 0.8848 - val_recall: 0.8857
Epoch 2/1000
10240/10268 [============================>.] - ETA: 0s - loss: 0.3347 - accuracy: 0.8818 - recall: 0.8830
Epoch 00002: val_recall did not improve from 0.94130
10268/10268 [==============================] - 3s 293us/sample - loss: 0.3350 - accuracy: 0.8820 - recall: 0.8831 - val_loss: 0.3351 - val_accuracy: 0.8826 - val_recall: 0.8828
Epoch 3/1000
10016/10268 [============================>.] - ETA: 0s - loss: 0.3357 - accuracy: 0.8819 - recall: 0.8834
Epoch 00003: val_recall did not improve from 0.94130
10268/10268 [==============================] - 3s 248us/sample - loss: 0.3355 - a

2567/2567 [==============================] - 0s 110us/sample - loss: 0.2366 - accuracy: 0.9048 - recall: 0.9062


Train on 10268 samples, validate on 5502 samples
Epoch 1/1000
10048/10268 [============================>.] - ETA: 0s - loss: 0.5235 - accuracy: 0.8701 - recall: 0.8736
Epoch 00001: val_recall did not improve from 0.94130
10268/10268 [==============================] - 3s 274us/sample - loss: 0.5208 - accuracy: 0.8703 - recall: 0.8736 - val_loss: 0.3967 - val_accuracy: 0.8821 - val_recall: 0.8846
Epoch 2/1000
10080/10268 [============================>.] - ETA: 0s - loss: 0.3267 - accuracy: 0.8865 - recall: 0.8861
Epoch 00002: val_recall did not improve from 0.94130
10268/10268 [==============================] - 2s 240us/sample - loss: 0.3266 - accuracy: 0.8864 - recall: 0.8862 - val_loss: 0.2785 - val_accuracy: 0.8867 - val_recall: 0.8837
Epoch 3/1000
10016/10268 [============================>.] - ETA: 0s - loss: 0.3074 - accuracy: 0.8907 - recall: 0.8908
Epoch 00003: val_recall did not improve from 0.94130
10268/10268 [==============================] - 3s 256us/sample - loss: 0.3073 - a

2567/2567 [==============================] - 0s 130us/sample - loss: 0.2780 - accuracy: 0.8954 - recall: 0.8960


Train on 10268 samples, validate on 5502 samples
Epoch 1/1000
10112/10268 [============================>.] - ETA: 0s - loss: 0.5785 - accuracy: 0.8616 - recall: 0.8662
Epoch 00001: val_recall did not improve from 0.94130
10268/10268 [==============================] - 3s 322us/sample - loss: 0.5742 - accuracy: 0.8619 - recall: 0.8665 - val_loss: 0.3146 - val_accuracy: 0.8797 - val_recall: 0.8810
Epoch 2/1000
10048/10268 [============================>.] - ETA: 0s - loss: 0.3515 - accuracy: 0.8811 - recall: 0.8813
Epoch 00002: val_recall did not improve from 0.94130
10268/10268 [==============================] - 3s 258us/sample - loss: 0.3501 - accuracy: 0.8812 - recall: 0.8813 - val_loss: 0.3325 - val_accuracy: 0.8678 - val_recall: 0.8674
Epoch 3/1000
10048/10268 [============================>.] - ETA: 0s - loss: 0.3107 - accuracy: 0.8869 - recall: 0.8864
Epoch 00003: val_recall did not improve from 0.94130
10268/10268 [==============================] - 3s 253us/sample - loss: 0.3106 - a

2567/2567 [==============================] - 0s 112us/sample - loss: 0.2714 - accuracy: 0.8987 - recall: 0.8993


Train on 10268 samples, validate on 5502 samples
Epoch 1/1000
10176/10268 [============================>.] - ETA: 0s - loss: 0.5554 - accuracy: 0.8575 - recall: 0.8574
Epoch 00001: val_recall did not improve from 0.94130
10268/10268 [==============================] - 3s 276us/sample - loss: 0.5524 - accuracy: 0.8580 - recall: 0.8580 - val_loss: 0.3729 - val_accuracy: 0.8797 - val_recall: 0.8815
Epoch 2/1000
10048/10268 [============================>.] - ETA: 0s - loss: 0.3541 - accuracy: 0.8806 - recall: 0.8793
Epoch 00002: val_recall did not improve from 0.94130
10268/10268 [==============================] - 2s 239us/sample - loss: 0.3538 - accuracy: 0.8803 - recall: 0.8791 - val_loss: 0.3450 - val_accuracy: 0.8654 - val_recall: 0.8777
Epoch 3/1000
10016/10268 [============================>.] - ETA: 0s - loss: 0.3169 - accuracy: 0.8837 - recall: 0.8848
Epoch 00003: val_recall did not improve from 0.94130
10268/10268 [==============================] - 2s 239us/sample - loss: 0.3155 - a

2567/2567 [==============================] - 0s 110us/sample - loss: 0.2709 - accuracy: 0.9083 - recall: 0.9020


Train on 10268 samples, validate on 5502 samples
Epoch 1/1000
10016/10268 [============================>.] - ETA: 0s - loss: 0.5639 - accuracy: 0.8698 - recall: 0.8699
Epoch 00001: val_recall did not improve from 0.94130
10268/10268 [==============================] - 3s 273us/sample - loss: 0.5566 - accuracy: 0.8705 - recall: 0.8706 - val_loss: 0.3174 - val_accuracy: 0.8847 - val_recall: 0.8884
Epoch 2/1000
10048/10268 [============================>.] - ETA: 0s - loss: 0.3104 - accuracy: 0.8901 - recall: 0.8901
Epoch 00002: val_recall did not improve from 0.94130
10268/10268 [==============================] - 2s 240us/sample - loss: 0.3090 - accuracy: 0.8903 - recall: 0.8903 - val_loss: 0.2714 - val_accuracy: 0.8912 - val_recall: 0.8938
Epoch 3/1000
10080/10268 [============================>.] - ETA: 0s - loss: 0.2858 - accuracy: 0.8946 - recall: 0.8956
Epoch 00003: val_recall did not improve from 0.94130
10268/10268 [==============================] - 2s 240us/sample - loss: 0.2880 - a

2567/2567 [==============================] - 0s 104us/sample - loss: 0.2486 - accuracy: 0.9011 - recall: 0.9006


Train on 10268 samples, validate on 5502 samples
Epoch 1/1000
10048/10268 [============================>.] - ETA: 0s - loss: 0.4977 - accuracy: 0.8715 - recall: 0.8686
Epoch 00001: val_recall did not improve from 0.94130
10268/10268 [==============================] - 3s 310us/sample - loss: 0.4939 - accuracy: 0.8723 - recall: 0.8693 - val_loss: 0.3040 - val_accuracy: 0.8845 - val_recall: 0.8868
Epoch 2/1000
10048/10268 [============================>.] - ETA: 0s - loss: 0.3341 - accuracy: 0.8890 - recall: 0.8890
Epoch 00002: val_recall did not improve from 0.94130
10268/10268 [==============================] - 3s 245us/sample - loss: 0.3339 - accuracy: 0.8890 - recall: 0.8890 - val_loss: 0.3015 - val_accuracy: 0.8869 - val_recall: 0.8847
Epoch 3/1000
 9952/10268 [============================>.] - ETA: 0s - loss: 0.2789 - accuracy: 0.8968 - recall: 0.8968
Epoch 00003: val_recall did not improve from 0.94130
10268/10268 [==============================] - 3s 255us/sample - loss: 0.2812 - a

2567/2567 [==============================] - 0s 99us/sample - loss: 0.2358 - accuracy: 0.9067 - recall: 0.9082


Train on 10268 samples, validate on 5502 samples
Epoch 1/1000
10112/10268 [============================>.] - ETA: 0s - loss: 0.5387 - accuracy: 0.8720 - recall: 0.8717
Epoch 00001: val_recall did not improve from 0.94130
10268/10268 [==============================] - 3s 251us/sample - loss: 0.5345 - accuracy: 0.8721 - recall: 0.8721 - val_loss: 0.3316 - val_accuracy: 0.8779 - val_recall: 0.9011
Epoch 2/1000
10176/10268 [============================>.] - ETA: 0s - loss: 0.3412 - accuracy: 0.8942 - recall: 0.8937
Epoch 00002: val_recall did not improve from 0.94130
10268/10268 [==============================] - 2s 218us/sample - loss: 0.3429 - accuracy: 0.8943 - recall: 0.8937 - val_loss: 0.2866 - val_accuracy: 0.8899 - val_recall: 0.8868
Epoch 3/1000
10016/10268 [============================>.] - ETA: 0s - loss: 0.3039 - accuracy: 0.8960 - recall: 0.8968
Epoch 00003: val_recall did not improve from 0.94130
10268/10268 [==============================] - 2s 223us/sample - loss: 0.3022 - a

2567/2567 [==============================] - 0s 107us/sample - loss: 0.2711 - accuracy: 0.8974 - recall: 0.8964


Train on 10268 samples, validate on 5502 samples
Epoch 1/1000
10080/10268 [============================>.] - ETA: 0s - loss: 0.5024 - accuracy: 0.8704 - recall: 0.8707
Epoch 00001: val_recall did not improve from 0.94130
10268/10268 [==============================] - 4s 382us/sample - loss: 0.5007 - accuracy: 0.8700 - recall: 0.8705 - val_loss: 0.3085 - val_accuracy: 0.8822 - val_recall: 0.8699
Epoch 2/1000
10176/10268 [============================>.] - ETA: 0s - loss: 0.3000 - accuracy: 0.8920 - recall: 0.8921
Epoch 00002: val_recall did not improve from 0.94130
10268/10268 [==============================] - 3s 248us/sample - loss: 0.2989 - accuracy: 0.8924 - recall: 0.8925 - val_loss: 0.2706 - val_accuracy: 0.8714 - val_recall: 0.8720
Epoch 3/1000
 9984/10268 [============================>.] - ETA: 0s - loss: 0.2765 - accuracy: 0.8917 - recall: 0.8914
Epoch 00003: val_recall did not improve from 0.94130
10268/10268 [==============================] - 2s 243us/sample - loss: 0.2770 - a

2567/2567 [==============================] - 0s 112us/sample - loss: 0.2336 - accuracy: 0.9077 - recall: 0.9078


Train on 10268 samples, validate on 5502 samples
Epoch 1/1000
10112/10268 [============================>.] - ETA: 0s - loss: 0.5571 - accuracy: 0.8625 - recall: 0.8648
Epoch 00001: val_recall did not improve from 0.94130
10268/10268 [==============================] - 3s 304us/sample - loss: 0.5526 - accuracy: 0.8632 - recall: 0.8654 - val_loss: 0.2941 - val_accuracy: 0.8796 - val_recall: 0.8813
Epoch 2/1000
10240/10268 [============================>.] - ETA: 0s - loss: 0.3047 - accuracy: 0.8869 - recall: 0.8887
Epoch 00002: val_recall did not improve from 0.94130
10268/10268 [==============================] - 3s 247us/sample - loss: 0.3049 - accuracy: 0.8869 - recall: 0.8887 - val_loss: 0.2919 - val_accuracy: 0.8783 - val_recall: 0.8808
Epoch 3/1000
10016/10268 [============================>.] - ETA: 0s - loss: 0.2974 - accuracy: 0.8905 - recall: 0.8913
Epoch 00003: val_recall did not improve from 0.94130
10268/10268 [==============================] - 2s 230us/sample - loss: 0.2959 - a

2567/2567 [==============================] - 0s 148us/sample - loss: 0.2188 - accuracy: 0.9131 - recall: 0.9128


Train on 10268 samples, validate on 5502 samples
Epoch 1/1000
10240/10268 [============================>.] - ETA: 0s - loss: 1.0105 - accuracy: 0.8632 - recall: 0.8712
Epoch 00001: val_recall did not improve from 0.94130
10268/10268 [==============================] - 3s 329us/sample - loss: 1.0091 - accuracy: 0.8634 - recall: 0.8715 - val_loss: 0.6493 - val_accuracy: 0.8707 - val_recall: 0.8980
Epoch 2/1000
10080/10268 [============================>.] - ETA: 0s - loss: 0.4984 - accuracy: 0.8704 - recall: 0.8675
Epoch 00002: val_recall did not improve from 0.94130
10268/10268 [==============================] - 3s 285us/sample - loss: 0.4957 - accuracy: 0.8707 - recall: 0.8677 - val_loss: 0.3442 - val_accuracy: 0.8852 - val_recall: 0.8810
Epoch 3/1000
10080/10268 [============================>.] - ETA: 0s - loss: 0.3456 - accuracy: 0.8834 - recall: 0.8830
Epoch 00003: val_recall did not improve from 0.94130
10268/10268 [==============================] - 3s 268us/sample - loss: 0.3473 - a

2567/2567 [==============================] - 0s 105us/sample - loss: 0.2913 - accuracy: 0.8925 - recall: 0.8987


Train on 10268 samples, validate on 5502 samples
Epoch 1/1000
10144/10268 [============================>.] - ETA: 0s - loss: 0.9548 - accuracy: 0.8587 - recall: 0.8550
Epoch 00001: val_recall did not improve from 0.94130
10268/10268 [==============================] - 3s 281us/sample - loss: 0.9468 - accuracy: 0.8591 - recall: 0.8556 - val_loss: 0.4566 - val_accuracy: 0.8708 - val_recall: 0.8999
Epoch 2/1000
10144/10268 [============================>.] - ETA: 0s - loss: 0.4172 - accuracy: 0.8698 - recall: 0.8692
Epoch 00002: val_recall did not improve from 0.94130
10268/10268 [==============================] - 3s 251us/sample - loss: 0.4171 - accuracy: 0.8697 - recall: 0.8691 - val_loss: 0.3071 - val_accuracy: 0.8763 - val_recall: 0.8838
Epoch 3/1000
10240/10268 [============================>.] - ETA: 0s - loss: 0.3302 - accuracy: 0.8819 - recall: 0.8795
Epoch 00003: val_recall did not improve from 0.94130
10268/10268 [==============================] - 3s 248us/sample - loss: 0.3305 - a

2567/2567 [==============================] - 0s 105us/sample - loss: 0.2389 - accuracy: 0.9055 - recall: 0.9070


Train on 10268 samples, validate on 5502 samples
Epoch 1/1000
10080/10268 [============================>.] - ETA: 0s - loss: 1.0040 - accuracy: 0.8585 - recall: 0.8585
Epoch 00001: val_recall did not improve from 0.94130
10268/10268 [==============================] - 3s 290us/sample - loss: 0.9955 - accuracy: 0.8587 - recall: 0.8590 - val_loss: 0.5027 - val_accuracy: 0.8642 - val_recall: 0.8813
Epoch 2/1000
10144/10268 [============================>.] - ETA: 0s - loss: 0.4956 - accuracy: 0.8706 - recall: 0.8701
Epoch 00002: val_recall did not improve from 0.94130
10268/10268 [==============================] - 3s 280us/sample - loss: 0.4956 - accuracy: 0.8706 - recall: 0.8700 - val_loss: 0.3590 - val_accuracy: 0.8549 - val_recall: 0.7994
Epoch 3/1000
10080/10268 [============================>.] - ETA: 0s - loss: 0.3448 - accuracy: 0.8866 - recall: 0.8867
Epoch 00003: val_recall did not improve from 0.94130
10268/10268 [==============================] - 3s 300us/sample - loss: 0.3436 - a

2567/2567 [==============================] - 0s 126us/sample - loss: 0.2998 - accuracy: 0.8921 - recall: 0.8925


Train on 10268 samples, validate on 5502 samples
Epoch 1/1000
10208/10268 [============================>.] - ETA: 0s - loss: 0.9548 - accuracy: 0.8566 - recall: 0.8566
Epoch 00001: val_recall did not improve from 0.94130
10268/10268 [==============================] - 4s 351us/sample - loss: 0.9517 - accuracy: 0.8569 - recall: 0.8570 - val_loss: 0.4534 - val_accuracy: 0.8640 - val_recall: 0.8382
Epoch 2/1000
10240/10268 [============================>.] - ETA: 0s - loss: 0.3992 - accuracy: 0.8749 - recall: 0.8733
Epoch 00002: val_recall did not improve from 0.94130
10268/10268 [==============================] - 3s 290us/sample - loss: 0.3988 - accuracy: 0.8750 - recall: 0.8735 - val_loss: 0.3386 - val_accuracy: 0.8838 - val_recall: 0.8828
Epoch 3/1000
10208/10268 [============================>.] - ETA: 0s - loss: 0.3316 - accuracy: 0.8851 - recall: 0.8827
Epoch 00003: val_recall did not improve from 0.94130
10268/10268 [==============================] - 3s 322us/sample - loss: 0.3314 - a

2567/2567 [==============================] - 0s 107us/sample - loss: 0.2406 - accuracy: 0.9067 - recall: 0.9074


Train on 10268 samples, validate on 5502 samples
Epoch 1/1000
10048/10268 [============================>.] - ETA: 0s - loss: 0.8884 - accuracy: 0.8513 - recall: 0.8506
Epoch 00001: val_recall did not improve from 0.94130
10268/10268 [==============================] - 3s 287us/sample - loss: 0.8827 - accuracy: 0.8510 - recall: 0.8503 - val_loss: 0.3905 - val_accuracy: 0.8621 - val_recall: 0.8939
Epoch 2/1000
10176/10268 [============================>.] - ETA: 0s - loss: 0.4484 - accuracy: 0.8685 - recall: 0.8692
Epoch 00002: val_recall did not improve from 0.94130
10268/10268 [==============================] - 3s 251us/sample - loss: 0.4492 - accuracy: 0.8680 - recall: 0.8691 - val_loss: 0.3343 - val_accuracy: 0.8669 - val_recall: 0.8597
Epoch 3/1000
10240/10268 [============================>.] - ETA: 0s - loss: 0.3440 - accuracy: 0.8798 - recall: 0.8755
Epoch 00003: val_recall did not improve from 0.94130
10268/10268 [==============================] - 3s 250us/sample - loss: 0.3438 - a

2567/2567 [==============================] - 0s 115us/sample - loss: 0.2342 - accuracy: 0.9135 - recall: 0.9140


Train on 10268 samples, validate on 5502 samples
Epoch 1/1000
10112/10268 [============================>.] - ETA: 0s - loss: 0.8176 - accuracy: 0.8512 - recall: 0.8525
Epoch 00001: val_recall did not improve from 0.94130
10268/10268 [==============================] - 3s 287us/sample - loss: 0.8149 - accuracy: 0.8501 - recall: 0.8507 - val_loss: 0.3823 - val_accuracy: 0.8352 - val_recall: 0.7950
Epoch 2/1000
10208/10268 [============================>.] - ETA: 0s - loss: 0.4724 - accuracy: 0.8604 - recall: 0.8551
Epoch 00002: val_recall did not improve from 0.94130
10268/10268 [==============================] - 2s 225us/sample - loss: 0.4725 - accuracy: 0.8605 - recall: 0.8555 - val_loss: 0.3562 - val_accuracy: 0.8774 - val_recall: 0.8902
Epoch 3/1000
10208/10268 [============================>.] - ETA: 0s - loss: 0.3623 - accuracy: 0.8723 - recall: 0.8724
Epoch 00003: val_recall did not improve from 0.94130
10268/10268 [==============================] - 3s 279us/sample - loss: 0.3625 - a

2567/2567 [==============================] - 0s 113us/sample - loss: 0.2722 - accuracy: 0.8966 - recall: 0.8987


Train on 10268 samples, validate on 5502 samples
Epoch 1/1000
10176/10268 [============================>.] - ETA: 0s - loss: 0.8787 - accuracy: 0.8576 - recall: 0.8541
Epoch 00001: val_recall did not improve from 0.94130
10268/10268 [==============================] - 3s 269us/sample - loss: 0.8762 - accuracy: 0.8573 - recall: 0.8541 - val_loss: 0.5080 - val_accuracy: 0.8608 - val_recall: 0.9173
Epoch 2/1000
10176/10268 [============================>.] - ETA: 0s - loss: 0.5157 - accuracy: 0.8674 - recall: 0.8729
Epoch 00002: val_recall did not improve from 0.94130
10268/10268 [==============================] - 3s 293us/sample - loss: 0.5152 - accuracy: 0.8673 - recall: 0.8727 - val_loss: 0.3378 - val_accuracy: 0.8733 - val_recall: 0.8844
Epoch 3/1000
10112/10268 [============================>.] - ETA: 0s - loss: 0.3867 - accuracy: 0.8706 - recall: 0.8691
Epoch 00003: val_recall did not improve from 0.94130
10268/10268 [==============================] - 3s 284us/sample - loss: 0.3862 - a

2567/2567 [==============================] - 0s 113us/sample - loss: 0.2612 - accuracy: 0.9020 - recall: 0.9028


Train on 10268 samples, validate on 5502 samples
Epoch 1/1000
10144/10268 [============================>.] - ETA: 0s - loss: 0.9201 - accuracy: 0.8550 - recall: 0.8556
Epoch 00001: val_recall did not improve from 0.94130
10268/10268 [==============================] - 3s 261us/sample - loss: 0.9176 - accuracy: 0.8554 - recall: 0.8564 - val_loss: 0.4654 - val_accuracy: 0.8402 - val_recall: 0.9235
Epoch 2/1000
10240/10268 [============================>.] - ETA: 0s - loss: 0.5347 - accuracy: 0.8601 - recall: 0.8621
Epoch 00002: val_recall did not improve from 0.94130
10268/10268 [==============================] - 2s 229us/sample - loss: 0.5340 - accuracy: 0.8601 - recall: 0.8620 - val_loss: 0.4196 - val_accuracy: 0.8061 - val_recall: 0.7252
Epoch 3/1000
10176/10268 [============================>.] - ETA: 0s - loss: 0.3970 - accuracy: 0.8671 - recall: 0.8753
Epoch 00003: val_recall did not improve from 0.94130
10268/10268 [==============================] - 2s 220us/sample - loss: 0.3956 - a

2567/2567 [==============================] - 0s 118us/sample - loss: 0.2870 - accuracy: 0.8851 - recall: 0.8914


Train on 10268 samples, validate on 5502 samples
Epoch 1/1000
10080/10268 [============================>.] - ETA: 0s - loss: 0.9648 - accuracy: 0.8562 - recall: 0.8551
Epoch 00001: val_recall did not improve from 0.94130
10268/10268 [==============================] - 3s 295us/sample - loss: 0.9552 - accuracy: 0.8561 - recall: 0.8552 - val_loss: 0.6328 - val_accuracy: 0.8304 - val_recall: 0.8266
Epoch 2/1000
10080/10268 [============================>.] - ETA: 0s - loss: 0.5490 - accuracy: 0.8583 - recall: 0.8624
Epoch 00002: val_recall did not improve from 0.94130
10268/10268 [==============================] - 2s 233us/sample - loss: 0.5469 - accuracy: 0.8585 - recall: 0.8629 - val_loss: 0.3574 - val_accuracy: 0.8789 - val_recall: 0.8897
Epoch 3/1000
10240/10268 [============================>.] - ETA: 0s - loss: 0.4082 - accuracy: 0.8694 - recall: 0.8760
Epoch 00003: val_recall did not improve from 0.94130
10268/10268 [==============================] - 2s 235us/sample - loss: 0.4076 - a

2567/2567 [==============================] - 0s 101us/sample - loss: 0.2716 - accuracy: 0.9067 - recall: 0.9074


Train on 10268 samples, validate on 5502 samples
Epoch 1/1000
 9984/10268 [============================>.] - ETA: 0s - loss: 0.8550 - accuracy: 0.8487 - recall: 0.8504
Epoch 00001: val_recall did not improve from 0.94130
10268/10268 [==============================] - 3s 258us/sample - loss: 0.8461 - accuracy: 0.8487 - recall: 0.8501 - val_loss: 0.3999 - val_accuracy: 0.8788 - val_recall: 0.8831
Epoch 2/1000
10016/10268 [============================>.] - ETA: 0s - loss: 0.4588 - accuracy: 0.8613 - recall: 0.8628
Epoch 00002: val_recall did not improve from 0.94130
10268/10268 [==============================] - 2s 221us/sample - loss: 0.4576 - accuracy: 0.8617 - recall: 0.8632 - val_loss: 0.3878 - val_accuracy: 0.8795 - val_recall: 0.8575
Epoch 3/1000
10176/10268 [============================>.] - ETA: 0s - loss: 0.3952 - accuracy: 0.8731 - recall: 0.8661
Epoch 00003: val_recall did not improve from 0.94130
10268/10268 [==============================] - 3s 289us/sample - loss: 0.3951 - a

2567/2567 [==============================] - 0s 109us/sample - loss: 0.2413 - accuracy: 0.9131 - recall: 0.9117


Train on 10268 samples, validate on 5502 samples
Epoch 1/1000
10240/10268 [============================>.] - ETA: 0s - loss: 1.0187 - accuracy: 0.8651 - recall: 0.8581
Epoch 00001: val_recall did not improve from 0.94130
10268/10268 [==============================] - 3s 321us/sample - loss: 1.0161 - accuracy: 0.8654 - recall: 0.8584 - val_loss: 0.4868 - val_accuracy: 0.8710 - val_recall: 0.8601
Epoch 2/1000
10144/10268 [============================>.] - ETA: 0s - loss: 0.4455 - accuracy: 0.8681 - recall: 0.8654
Epoch 00002: val_recall did not improve from 0.94130
10268/10268 [==============================] - 3s 274us/sample - loss: 0.4435 - accuracy: 0.8679 - recall: 0.8652 - val_loss: 0.3340 - val_accuracy: 0.8762 - val_recall: 0.8786
Epoch 3/1000
10112/10268 [============================>.] - ETA: 0s - loss: 0.3249 - accuracy: 0.8825 - recall: 0.8810
Epoch 00003: val_recall did not improve from 0.94130
10268/10268 [==============================] - 3s 285us/sample - loss: 0.3244 - a

2567/2567 [==============================] - 0s 115us/sample - loss: 0.2456 - accuracy: 0.8975 - recall: 0.8975


Train on 10268 samples, validate on 5502 samples
Epoch 1/1000
10048/10268 [============================>.] - ETA: 0s - loss: 0.9414 - accuracy: 0.8562 - recall: 0.8573
Epoch 00001: val_recall did not improve from 0.94130
10268/10268 [==============================] - 3s 289us/sample - loss: 0.9287 - accuracy: 0.8575 - recall: 0.8584 - val_loss: 0.4431 - val_accuracy: 0.8564 - val_recall: 0.8573
Epoch 2/1000
10016/10268 [============================>.] - ETA: 0s - loss: 0.4436 - accuracy: 0.8760 - recall: 0.8746- ETA: 0s - loss: 0.4957 - accuracy: 0.8682
Epoch 00002: val_recall did not improve from 0.94130
10268/10268 [==============================] - 2s 242us/sample - loss: 0.4408 - accuracy: 0.8763 - recall: 0.8752 - val_loss: 0.3138 - val_accuracy: 0.8824 - val_recall: 0.8969
Epoch 3/1000
10208/10268 [============================>.] - ETA: 0s - loss: 0.3532 - accuracy: 0.8838 - recall: 0.8827
Epoch 00003: val_recall did not improve from 0.94130
10268/10268 [=========================

2567/2567 [==============================] - 0s 138us/sample - loss: 0.2611 - accuracy: 0.9007 - recall: 0.9016


Train on 10268 samples, validate on 5502 samples
Epoch 1/1000
10176/10268 [============================>.] - ETA: 0s - loss: 1.1111 - accuracy: 0.8627 - recall: 0.8600
Epoch 00001: val_recall did not improve from 0.94130
10268/10268 [==============================] - 3s 325us/sample - loss: 1.1049 - accuracy: 0.8634 - recall: 0.8606 - val_loss: 0.8482 - val_accuracy: 0.8765 - val_recall: 0.8671
Epoch 2/1000
10080/10268 [============================>.] - ETA: 0s - loss: 0.6425 - accuracy: 0.8644 - recall: 0.8628
Epoch 00002: val_recall did not improve from 0.94130
10268/10268 [==============================] - 3s 244us/sample - loss: 0.6389 - accuracy: 0.8643 - recall: 0.8624 - val_loss: 0.4256 - val_accuracy: 0.8627 - val_recall: 0.8426
Epoch 3/1000
10144/10268 [============================>.] - ETA: 0s - loss: 0.3985 - accuracy: 0.8790 - recall: 0.8801
Epoch 00003: val_recall did not improve from 0.94130
10268/10268 [==============================] - 3s 246us/sample - loss: 0.3985 - a

2567/2567 [==============================] - 0s 113us/sample - loss: 0.2671 - accuracy: 0.8849 - recall: 0.8852


Train on 10268 samples, validate on 5502 samples
Epoch 1/1000
10112/10268 [============================>.] - ETA: 0s - loss: 1.0419 - accuracy: 0.8598 - recall: 0.8677
Epoch 00001: val_recall did not improve from 0.94130
10268/10268 [==============================] - 3s 316us/sample - loss: 1.0332 - accuracy: 0.8602 - recall: 0.8673 - val_loss: 0.7690 - val_accuracy: 0.8487 - val_recall: 0.8103
Epoch 2/1000
10016/10268 [============================>.] - ETA: 0s - loss: 0.4781 - accuracy: 0.8656 - recall: 0.8626
Epoch 00002: val_recall did not improve from 0.94130
10268/10268 [==============================] - 3s 259us/sample - loss: 0.4730 - accuracy: 0.8663 - recall: 0.8634 - val_loss: 0.3377 - val_accuracy: 0.8775 - val_recall: 0.8750
Epoch 3/1000
10240/10268 [============================>.] - ETA: 0s - loss: 0.3196 - accuracy: 0.8853 - recall: 0.8845
Epoch 00003: val_recall did not improve from 0.94130
10268/10268 [==============================] - 3s 268us/sample - loss: 0.3193 - a

2567/2567 [==============================] - 0s 103us/sample - loss: 0.2464 - accuracy: 0.9042 - recall: 0.9043


Train on 10268 samples, validate on 5502 samples
Epoch 1/1000
10112/10268 [============================>.] - ETA: 0s - loss: 1.0672 - accuracy: 0.8597 - recall: 0.8658
Epoch 00001: val_recall did not improve from 0.94130
10268/10268 [==============================] - 3s 283us/sample - loss: 1.0663 - accuracy: 0.8598 - recall: 0.8665 - val_loss: 0.7982 - val_accuracy: 0.8478 - val_recall: 0.8581
Epoch 2/1000
10144/10268 [============================>.] - ETA: 0s - loss: 0.5726 - accuracy: 0.8637 - recall: 0.8629
Epoch 00002: val_recall did not improve from 0.94130
10268/10268 [==============================] - 3s 246us/sample - loss: 0.5704 - accuracy: 0.8638 - recall: 0.8626 - val_loss: 0.3466 - val_accuracy: 0.8705 - val_recall: 0.8275
Epoch 3/1000
10080/10268 [============================>.] - ETA: 0s - loss: 0.3727 - accuracy: 0.8789 - recall: 0.8789
Epoch 00003: val_recall did not improve from 0.94130
10268/10268 [==============================] - 3s 257us/sample - loss: 0.3705 - a

2567/2567 [==============================] - 0s 111us/sample - loss: 0.2418 - accuracy: 0.9139 - recall: 0.9151


Train on 10268 samples, validate on 5502 samples
Epoch 1/1000
10176/10268 [============================>.] - ETA: 0s - loss: 0.6886 - accuracy: 0.8572 - recall: 0.8561
Epoch 00001: val_recall did not improve from 0.94130
10268/10268 [==============================] - 3s 246us/sample - loss: 0.6929 - accuracy: 0.8566 - recall: 0.8557 - val_loss: 0.4818 - val_accuracy: 0.8770 - val_recall: 0.8853
Epoch 2/1000
 9952/10268 [============================>.] - ETA: 0s - loss: 0.3877 - accuracy: 0.8821 - recall: 0.8827
Epoch 00002: val_recall did not improve from 0.94130
10268/10268 [==============================] - 2s 214us/sample - loss: 0.3828 - accuracy: 0.8828 - recall: 0.8835 - val_loss: 0.3484 - val_accuracy: 0.8888 - val_recall: 0.8864
Epoch 3/1000
10048/10268 [============================>.] - ETA: 0s - loss: 0.3312 - accuracy: 0.8873 - recall: 0.8884
Epoch 00003: val_recall did not improve from 0.94130
10268/10268 [==============================] - 2s 238us/sample - loss: 0.3342 - a

2567/2567 [==============================] - 0s 99us/sample - loss: 0.2625 - accuracy: 0.8991 - recall: 0.9029


Train on 10268 samples, validate on 5502 samples
Epoch 1/1000
10144/10268 [============================>.] - ETA: 0s - loss: 0.6798 - accuracy: 0.8643 - recall: 0.8648
Epoch 00001: val_recall did not improve from 0.94130
10268/10268 [==============================] - 3s 258us/sample - loss: 0.6744 - accuracy: 0.8651 - recall: 0.8658 - val_loss: 0.3552 - val_accuracy: 0.8827 - val_recall: 0.8868
Epoch 2/1000
10016/10268 [============================>.] - ETA: 0s - loss: 0.3691 - accuracy: 0.8814 - recall: 0.8839
Epoch 00002: val_recall did not improve from 0.94130
10268/10268 [==============================] - 2s 214us/sample - loss: 0.3699 - accuracy: 0.8813 - recall: 0.8839 - val_loss: 0.3020 - val_accuracy: 0.8668 - val_recall: 0.8693
Epoch 3/1000
10208/10268 [============================>.] - ETA: 0s - loss: 0.3235 - accuracy: 0.8868 - recall: 0.8870
Epoch 00003: val_recall did not improve from 0.94130
10268/10268 [==============================] - 2s 217us/sample - loss: 0.3232 - a

2567/2567 [==============================] - 0s 98us/sample - loss: 0.2464 - accuracy: 0.8931 - recall: 0.8947


Train on 10268 samples, validate on 5502 samples
Epoch 1/1000
10112/10268 [============================>.] - ETA: 0s - loss: 0.6283 - accuracy: 0.8674 - recall: 0.8633
Epoch 00001: val_recall did not improve from 0.94130
10268/10268 [==============================] - 3s 248us/sample - loss: 0.6240 - accuracy: 0.8676 - recall: 0.8636 - val_loss: 0.3118 - val_accuracy: 0.8745 - val_recall: 0.8787
Epoch 2/1000
 9952/10268 [============================>.] - ETA: 0s - loss: 0.3515 - accuracy: 0.8875 - recall: 0.8882
Epoch 00002: val_recall did not improve from 0.94130
10268/10268 [==============================] - 2s 214us/sample - loss: 0.3496 - accuracy: 0.8877 - recall: 0.8883 - val_loss: 0.2941 - val_accuracy: 0.8690 - val_recall: 0.8666
Epoch 3/1000
10240/10268 [============================>.] - ETA: 0s - loss: 0.3097 - accuracy: 0.8923 - recall: 0.8884
Epoch 00003: val_recall did not improve from 0.94130
10268/10268 [==============================] - 2s 215us/sample - loss: 0.3096 - a

2567/2567 [==============================] - 0s 96us/sample - loss: 0.2820 - accuracy: 0.8960 - recall: 0.8968


Train on 10268 samples, validate on 5502 samples
Epoch 1/1000
10144/10268 [============================>.] - ETA: 0s - loss: 0.7966 - accuracy: 0.8540 - recall: 0.8498
Epoch 00001: val_recall did not improve from 0.94130
10268/10268 [==============================] - 3s 250us/sample - loss: 0.7920 - accuracy: 0.8541 - recall: 0.8498 - val_loss: 0.4453 - val_accuracy: 0.8755 - val_recall: 0.8712
Epoch 2/1000
10208/10268 [============================>.] - ETA: 0s - loss: 0.4146 - accuracy: 0.8750 - recall: 0.8732
Epoch 00002: val_recall did not improve from 0.94130
10268/10268 [==============================] - 2s 217us/sample - loss: 0.4180 - accuracy: 0.8749 - recall: 0.8731 - val_loss: 0.4019 - val_accuracy: 0.8879 - val_recall: 0.8817
Epoch 3/1000
10208/10268 [============================>.] - ETA: 0s - loss: 0.3344 - accuracy: 0.8874 - recall: 0.8852
Epoch 00003: val_recall did not improve from 0.94130
10268/10268 [==============================] - 2s 220us/sample - loss: 0.3334 - a

2567/2567 [==============================] - 0s 97us/sample - loss: 0.2621 - accuracy: 0.9030 - recall: 0.9039


Train on 10268 samples, validate on 5502 samples
Epoch 1/1000
10176/10268 [============================>.] - ETA: 0s - loss: 0.7569 - accuracy: 0.8531 - recall: 0.8562
Epoch 00001: val_recall did not improve from 0.94130
10268/10268 [==============================] - 3s 247us/sample - loss: 0.7538 - accuracy: 0.8535 - recall: 0.8564 - val_loss: 0.4450 - val_accuracy: 0.8594 - val_recall: 0.8441
Epoch 2/1000
 9952/10268 [============================>.] - ETA: 0s - loss: 0.4433 - accuracy: 0.8717 - recall: 0.8687
Epoch 00002: val_recall did not improve from 0.94130
10268/10268 [==============================] - 2s 215us/sample - loss: 0.4444 - accuracy: 0.8720 - recall: 0.8696 - val_loss: 0.3676 - val_accuracy: 0.8513 - val_recall: 0.7727
Epoch 3/1000
10208/10268 [============================>.] - ETA: 0s - loss: 0.3404 - accuracy: 0.8844 - recall: 0.8766
Epoch 00003: val_recall did not improve from 0.94130
10268/10268 [==============================] - 2s 216us/sample - loss: 0.3406 - a

2567/2567 [==============================] - 0s 98us/sample - loss: 0.2246 - accuracy: 0.9118 - recall: 0.9097


RuntimeError: Cannot clone object <tensorflow.python.keras.wrappers.scikit_learn.KerasClassifier object at 0x000002CA634CEB00>, as the constructor either does not set or modifies parameter dropout3

In [38]:
gs.best_params_

{'epochs': 1000,
 'dropout3': 0.4,
 'dropout2': 0.0,
 'dropout1': 0.0,
 'dense_dims3': 512,
 'dense_dims2': 128,
 'dense_dims1': 512}

### using the best model for evaluation

In [42]:
best_model = build_model(optimizer='adam',
     dropout3 = 0.4,
     dropout2 = 0.0,
     dropout1 = 0.0,
     dense_dims3 = 512,
     dense_dims2 = 128,
     dense_dims1 = 512)

In [43]:
best_model.fit(X_train,y_train,verbose=1,
        validation_data=(x_validation, y_validation),callbacks = callbacks, epochs = 1000)

Train on 12835 samples, validate on 5502 samples
Epoch 1/1000
12672/12835 [============================>.] - ETA: 0s - loss: 0.6020 - accuracy: 0.8718 - recall: 0.8707
Epoch 00001: val_recall did not improve from 0.94130
12835/12835 [==============================] - 3s 255us/sample - loss: 0.5974 - accuracy: 0.8723 - recall: 0.8716 - val_loss: 0.3270 - val_accuracy: 0.8866 - val_recall: 0.8867
Epoch 2/1000
12768/12835 [============================>.] - ETA: 0s - loss: 0.3389 - accuracy: 0.8892 - recall: 0.8909
Epoch 00002: val_recall did not improve from 0.94130
12835/12835 [==============================] - 3s 225us/sample - loss: 0.3390 - accuracy: 0.8893 - recall: 0.8912 - val_loss: 0.2636 - val_accuracy: 0.8930 - val_recall: 0.8924
Epoch 3/1000
12608/12835 [============================>.] - ETA: 0s - loss: 0.2851 - accuracy: 0.8974 - recall: 0.8971
Epoch 00003: val_recall did not improve from 0.94130
12835/12835 [==============================] - 3s 217us/sample - loss: 0.2853 - a

In [44]:
best_model.evaluate(x_validation, y_validation)

5502/1 [================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================

[0.2617971995762156, 0.8952199, 0.89551836]

__The best model from grid search gives a 89% recall on train and evaluation, there was a model with 96% recall but I have to do some research to fix the issue with randomsearchcv and keras.__